File: Day241/shortestPathVisitingAllNodes.java
package Day241;

import java.util.Queue;
import java.util.LinkedList;
import java.util.Arrays;

public class shortestPathVisitingAllNodes {

    public static int shortestPathLength(int[][] graph) {
        int answer = Integer.MAX_VALUE;
        int n = graph.length;
        int dp[][] = new int[1<<n][n];
        for (int i=0; i<1<<n; i++) {
            Arrays.fill(dp[i], Integer.MAX_VALUE);
        }
        for (int i=0; i<n; i++) {
            dp[1<<i][i] = 0;
        }
        Queue<int[]> queue = new LinkedList<>();
        
        for (int i=0; i<n; i++) {
            queue.offer(new int[]{1<<i, i});
        }
        while (!queue.isEmpty()){
            int temp[] = queue.poll();
            int mask = temp[0];
            int u = temp[1];
            for (int v : graph[u]) {
                int nm = mask | (1 <<v);
                if (dp[nm][v] == Integer.MAX_VALUE) {
                    dp[nm][v] = dp[mask][u] + 1;
                    queue.offer(new int[]{nm, v});
                }
            }
        }
        for (int i=0; i<n; i++) {
            answer = Math.min(answer, dp[(1<<n)-1][i]);
        }
        return answer;
    }

    public static void main(String[] args) {
        int arr1[][] = {{1,2,3},{0},{0},{0}};
        int arr2[][] = {{1},{0,2,4},{1,3,4},{2},{1,2}};
        System.out.println(shortestPathLength(arr1));
        System.out.println(shortestPathLength(arr2));
    }
}

File: Day241/shortestPathVisitingAllNodes.go
package main

import (
	"fmt"
	"math"
)

func min(a int , b int ) int {
	if a < b {
		return a
	}
	return b
}

func shortestPathLength(graph [][]int) int {
	answer := math.MaxInt
	n := len(graph)
	dp := make([][]int, 1<<n)
	for i:=0; i<1<<n; i++ {
		dp[i] = make([]int, n)
		for j:=0; j<n; j++ {
			dp[i][j] = math.MaxInt
		}
	}	
	for i:=0; i<n; i++ {
		dp[1<<i][i] = 0
	}
	queue := [][]int{}
	for i:=0; i<n; i++ {
		queue = append(queue, []int{1<<i, i})
	}
	for len(queue) > 0 {
		temp := queue[0]
		queue = queue[1:]
		mask := temp[0]
		u := temp[1]
		for _,v :=range graph[u] {
			nm := mask |  (1<<v)
			if dp[nm][v] == math.MaxInt {
				dp[nm][v] = dp[mask][u] + 1
				queue = append(queue, []int{nm, v})
			}
		}
	}
	for i:=0; i<n; i++ {
		answer = min(answer, dp[(1<<n)-1] [i])
	}
	return answer
}

func main() {
	arr1 := [][]int{{1,2,3},{0},{0},{0}}
	arr2 := [][]int{{1},{0,2,4},{1,3,4},{2},{1,2}}
	fmt.Println(shortestPathLength(arr1))
	fmt.Println(shortestPathLength(arr2))
}

